crates/neon/src/types_impl/buffer/types.rs

```diff
@@ -0,0 +1,783 @@
+    /// # fn f(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    /// let buf: Handle<JsArrayBuffer> = cx.argument(0)?;
+    /// let region = buf.region::<u32>(64, 8);
+    /// println!("offset={}, len={}, size={}", region.offset(), region.len(), region.size());
+    /// # Ok(cx.undefined())
+    /// # }
+    /// ```
+    ///
+    /// See the [`Region`](Region) documentation for more information.
+    pub fn region<T: Binary>(&self, offset: usize, len: usize) -> Region<'cx, T> {
+        Region {
+            buffer: *self,
+            offset,
+            len,
+            phantom: PhantomData,
+        }
+    }
+}
+
+unsafe impl TransparentNoCopyWrapper for JsArrayBuffer {
+    type Inner = raw::Local;
+
+    fn into_inner(self) -> Self::Inner {
+        self.0
+    }
+}
+
+impl Managed for JsArrayBuffer {
+    fn to_raw(&self) -> raw::Local {
+        self.0
+    }
+
+    fn from_raw(_env: Env, h: raw::Local) -> Self {
+        Self(h)
+    }
+}
+
+impl ValueInternal for JsArrayBuffer {
+    fn name() -> String {
+        "JsArrayBuffer".to_string()
+    }
+
+    fn is_typeof<Other: Value>(env: Env, other: &Other) -> bool {
+        unsafe { sys::tag::is_arraybuffer(env.to_raw(), other.to_raw()) }
+    }
+}
+
+impl Value for JsArrayBuffer {}
+
+impl Object for JsArrayBuffer {}
+
+impl private::Sealed for JsArrayBuffer {}
+
+impl TypedArray for JsArrayBuffer {
+    type Item = u8;
+
+    fn as_slice<'cx, 'a, C>(&self, cx: &'a C) -> &'a [Self::Item]
+    where
+        C: Context<'cx>,
+    {
+        unsafe { sys::arraybuffer::as_mut_slice(cx.env().to_raw(), self.to_raw()) }
+    }
+
+    fn as_mut_slice<'cx, 'a, C>(&mut self, cx: &'a mut C) -> &'a mut [Self::Item]
+    where
+        C: Context<'cx>,
+    {
+        unsafe { sys::arraybuffer::as_mut_slice(cx.env().to_raw(), self.to_raw()) }
+    }
+
+    fn try_borrow<'cx, 'a, C>(&self, lock: &'a Lock<C>) -> Result<Ref<'a, Self::Item>, BorrowError>
+    where
+        C: Context<'cx>,
+    {
+        // The borrowed data must be guarded by `Ledger` before returning
+        Ledger::try_borrow(&lock.ledger, unsafe {
+            sys::arraybuffer::as_mut_slice(lock.cx.env().to_raw(), self.to_raw())
+        })
+    }
+
+    fn try_borrow_mut<'cx, 'a, C>(
+        &mut self,
+        lock: &'a Lock<C>,
+    ) -> Result<RefMut<'a, Self::Item>, BorrowError>
+    where
+        C: Context<'cx>,
+    {
+        // The borrowed data must be guarded by `Ledger` before returning
+        Ledger::try_borrow_mut(&lock.ledger, unsafe {
+            sys::arraybuffer::as_mut_slice(lock.cx.env().to_raw(), self.to_raw())
+        })
+    }
+
+    fn size<'cx, C: Context<'cx>>(&self, cx: &mut C) -> usize {
+        unsafe { sys::arraybuffer::size(cx.env().to_raw(), self.to_raw()) }
+    }
+}
+
+/// A marker trait for all possible element types of binary buffers.
+///
+/// This trait can only be implemented within the Neon library.
+pub trait Binary: private::Sealed + Copy + std::fmt::Debug {
```

Comment:
Since these bounds are only required for the `PhantomData<T>` field of `JsTypedArrayInner`, I can reduce the requirements from `Debug + Copy` to just `Clone` by manually implementing `Debug`.

I'm a little reluctant to remove the `Clone` super type constraint and propagate it into the declarations, since it pollutes all the impl declarations everywhere. This means a bunch of methods just disappear when your `Binary` type happens not to be `Clone`, which just seems like `JsTypedArray` just doesn't really work as expected unless your `Binary` type is also `Clone`. I feel like I'd rather just express that it's a requirement of the trait.

It is an annoying requirement of the Rust compiler, though. It's frustrating that you can't declare `impl<T: Binary> Copy for JsTypedArrayInner<T> {}` if `T` isn't `Clone`, even though it's only used as a phantom type. (I wonder if that's what [this issue](https://github.com/rust-lang/rust/issues/50980) is about?)

But in the end it seems pretty reasonable to require a `Binary` type to be `Clone`, since binary data is just sequences of bits. What do you think of leaving it at `trait Binary: Sealed + Clone`?

Reply:
That makes sense to me. Thanks for the explanation!